bpf/bpf_host.c

```diff
@@ -1391,6 +1426,24 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
 	}
 #endif
 
+#if defined(ENABLE_IPSEC) && defined(ENABLE_IPV4) && defined(TUNNEL_MODE)
+	/* Determine if this is overlay traffic that should be recirculated
+	 * to the stack for XFRM encryption.
+	 */
+	ret = do_encrypt_overlay(ctx, proto);
+	if (ret == CTX_ACT_REDIRECT) {
+		/* we are redirecting back into the stack, so TRACE_TO_STACK
+		 * for tracepoint
+		 */
+		send_trace_notify(ctx, TRACE_TO_STACK, 0, 0, 0,
+				  0, TRACE_REASON_ENCRYPT_OVERLAY, 0);
```

Comment:
Two points here:
- I don't know if you've considered the meaning of all these `0`s, but I suspect that we're avoiding providing debugging information about the flow here, which will make this harder to debug if a user tries to figure out what's going on based on this trace message. Ideally very few of these `0`s should be `0`s, we can provide the real information about the packet at this point in time.
- I think we need to consider how many of these messages we want to generate and how this integrates with `monitor-aggregation`. If this is part of encrypt hotpath then it could be emitted for every packet, forcing Cilium to spend a bunch of CPU processing these events.

Reply:
This maybe a bit of a unique case, since when we get here we have vxlan encap'd traffic already. 

Given our last convo, we are going to have to pull the source information out to rewrite the VNI, so we can fill in the src ID, however, I'm not sure if filling in the dest_id is worth the effort, as we'd need to sniff into the vxlan traffic for the dest and do an additional map lookup. 

re: `monitor-aggregation` its a good question, this tracepoint is right before we redirect the packet back into the interface and route it toward XFRM so it can do the encryption. What is our stance for traces like this? Aren't all our traces susceptible to hot paths if the application starts bursting data?

Reply:
> Aren't all our traces susceptible to hot paths if the application starts bursting data?

Yep, and it's already a problem for CPU usage at high packet rates; I wouldn't want to extend this further with additional trace events that have no way to opt-out. There's been some PRs recently to try to tackle some of this, so maybe it's worth poking around in the recent hubble/monitor PRs and raise some discussion about the right way to introduce this visibility while accounting for the cost (or allowing it to be disabled).

Reply:
What Louis did here is hide those traces whenever monitor aggregation based on CT is enabled (cf. [`lib/trace.h`](https://github.com/cilium/cilium/blob/6b98a0b210ea2eae59e2eeb399aad3ce121f3caf/bpf/lib/trace.h#L174-L181)). So by default those traces won't be emitted. It matches Ryan's recent changes in https://github.com/cilium/cilium/pull/31015/commits/5fc9dd75d42fcf9214ebb8bd408ee91063f3c865 and https://github.com/cilium/cilium/pull/29616/commits/897b8edd9c495b29cc3344fd7fa300adda4b1b5b.

Not to say we shouldn't fill the contextual information (ex. security identity) when we have it.

Reply:
Something else I didn't notice the first time... is that every other `CTX_ACT_REDIRECT` path in this function *avoids* calling `send_trace_notify()`. I don't know if there's a specific reason for this or if it's an oversight, because the code structure doesn't make this clear. Maybe that's just a general datapath cleanup thing to follow up on.

Reply:
Yup make sense. 

I personally feel like a trace notification is really valuable here. The packet-path for encrypted overlay is kinda complex IMO, so knowing that traces which show the major leg of the flow is important IMO.

Reply:
I think we landed on keeping the trace as it makes sense since we are handing the packet off to the stack. So going to resolve. If this is not the case please unresolve and ping me.